certora/specs/DoubleEndedQueue.spec

```diff
@@ -0,0 +1,408 @@
+
+    pushBack@withrevert(value);
+    bool success = !lastReverted;
+
+    bytes32 backAfter = back();
+
+    // liveness
+    assert success, "never reverts";
+
+    // effect
+    assert success => backAfter == value, "back set to value";
+
+    // no side effect
+    assert backBefore == backAfter <=> backBefore == value, "back doesn't change only if it was already set";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: pushBack preserves the previous values in the queue                                                           │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule pushBackConsistency {
+    require boundedQueue();
+
+    uint256 key;
+    bytes32 beforeAt = at_(key);
+
+    bytes32 value;
+    pushBack@withrevert(value);
+
+    bytes32 afterAt = at_(key);
+
+    assert afterAt == beforeAt, "data is preserved";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: popFront removes an element from the beginning of the queue                                   │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popFront {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    bool emptyBefore = empty();
+    bytes32 frontBefore = front@withrevert();
+
+    bytes32 popped = popFront@withrevert();
+    bool success = !lastReverted;
+
+    // liveness
+    assert success <=> !emptyBefore, "only fails if it's empty";
+
+    // effect
+    assert success => frontBefore == popped, "previous front is returned";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: front is previous at(1) after calling popFront                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popFrontExpected {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+    
+    bytes32 nextFront = at_(1);
+    popFront@withrevert();
+
+    assert !lastReverted <=> front() == nextFront, "front is adjusted";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: at(x) is preserved and offset to at(x - 1) after calling popFront                                             |
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popFrontConsistency {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    uint256 key;
+    bytes32 before = at_(key);
```

Comment:
Note that the `at_` call on line 227 is not `@withrevert` ... so it forces the queue to be non empty, which means the `popFront` will never revert.

Also, the `at_` on line 232 adds additional contraints to the rule (queue is not empty after the `popFront`)

Reply:
I think I lost this reference by updating so I'm not sure where it's the suggestion pointing to, but I'm assuming is for `popBackConsistency` and `pushFrontConsistency`, I think it's not a problem anymore after adding your suggestion (eg. attempting to check `at_@wtihRevert` after popping)